server/base/src/main/java/org/apache/accumulo/server/GarbageCollectionLogger.java

```diff
@@ -22,109 +22,113 @@
 
-      if (mem != lastMemorySize) {
+      if (time - pt != 0) {
         sawChange = true;
       }
 
-      String sign = "+";
-      if (mem - lastMemorySize <= 0) {
-        sign = "";
+      long increaseInCollectionTime = time - pt;
+      sb.append(String.format(" %s=%,.2f(+%,.2f) secs", gcBean.getName(), time / 1000.0,
+          increaseInCollectionTime / 1000.0));
+      maxIncreaseInCollectionTime = Math.max(increaseInCollectionTime, maxIncreaseInCollectionTime);
+      prevGcTime.put(gcBean.getName(), time);
+    }
+
+    long mem = rt.freeMemory();
+    if (maxIncreaseInCollectionTime == 0) {
+      gcTimeIncreasedCount = 0;
+    } else {
+      gcTimeIncreasedCount++;
+      if (gcTimeIncreasedCount > 3 && mem < rt.maxMemory() * 0.05) {
+        log.warn("Running low on memory");
+        gcTimeIncreasedCount = 0;
       }
+    }
+
+    if (mem != lastMemorySize) {
+      sawChange = true;
+    }
 
-      sb.append(String.format(" freemem=%,d(%s%,d) totalmem=%,d", mem, sign, (mem - lastMemorySize),
-          rt.totalMemory()));
+    String sign = "+";
+    if (mem - lastMemorySize <= 0) {
+      sign = "";
+    }
 
-      if (sawChange) {
-        log.debug(sb.toString());
-      }
+    sb.append(String.format(" freemem=%,d(%s%,d) totalmem=%,d", mem, sign, (mem - lastMemorySize),
+        rt.totalMemory()));
 
-      final long keepAliveTimeout = conf.getTimeInMillis(Property.INSTANCE_ZK_TIMEOUT);
-      if (lastMemoryCheckTime > 0 && lastMemoryCheckTime < now) {
-        final long diff = now - lastMemoryCheckTime;
-        if (diff > keepAliveTimeout + 1000) {
-          log.warn(String.format(
-              "GC pause checker not called in a timely"
-                  + " fashion. Expected every %.1f seconds but was %.1f seconds since last check",
-              keepAliveTimeout / 1000., diff / 1000.));
-        }
-        lastMemoryCheckTime = now;
-        return;
-      }
+    if (sawChange) {
+      log.debug(sb.toString());
+    }
 
-      if (maxIncreaseInCollectionTime > keepAliveTimeout) {
-        Halt.halt("Garbage collection may be interfering with lock keep-alive.  Halting.", -1);
+    final long keepAliveTimeout =
+        context.getConfiguration().getTimeInMillis(Property.INSTANCE_ZK_TIMEOUT);
+    if (lastMemoryCheckTime > 0 && lastMemoryCheckTime < now) {
+      final long diff = now - lastMemoryCheckTime;
+      if (diff > keepAliveTimeout + 1000) {
+        log.warn(String.format(
+            "GC pause checker not called in a timely"
+                + " fashion. Expected every %.1f seconds but was %.1f seconds since last check",
+            keepAliveTimeout / 1000., diff / 1000.));
       }
-
-      lastMemorySize = mem;
       lastMemoryCheckTime = now;
-    } finally {
-      memCheckTimeLock.unlock();
+      return;
     }
-  }
 
+    if (maxIncreaseInCollectionTime > keepAliveTimeout) {
+      Halt.halt("Garbage collection may be interfering with lock keep-alive.  Halting.", -1);
```

Comment:
I'm wondering if we need to still preemptively Halt the server since we have better exception handling in background threads now. If an OOME due to GC Overhead is thrown in a background thread it will halt the VM.

Reply:
I also wonder if we still need this halt for a number of reasons.

Reply:
I'd be happy to remove it - maybe as a separate issue?  Do you have test that demonstrates that it is not needed?

Reply:
No, just thinking out loud.